### DISCUSSION TITLE
Listener for the "send" or wait before sending another command?

### DISCUSSION THREAD

**Discussion by jwjkim** [migueldeicaza/SwiftTerm#355] [Q&A]
I'm using the below function to send bash commands and it's working, but the problem is I want to send multiple commands once at a time. Is there a way for me to listen/wait until the first `send` is complete before sending another?

  /**
   * Sends the specified string encoded at utf8 to the program running under the terminal emulator
   * - Parameter txt: the string to send to the client
   */
  public func send (txt: String) {
      let array = [UInt8] (txt.utf8)
      send (data: array[...])
  }

**Comment by migueldeicaza** [comment]
You have control over the send infrastructure, so you can detect that condition in your implementation.

**Reply by jwjkim** [reply]
Hi, I've been trying to figure out how I can do this but I'm not seeing any listener I can use or a trigger that will notify that the send command finished. Could you explain how I can control the send? 

This is the structure I have to use the LocalProcessTerminalView so from my SwiftUI app, I'm only using the sendCommand. Is there something I can add to the sendCommand function?

Thank you!

````
import Foundation
import SwiftTerm
import SwiftUI

struct TerminalView: NSViewRepresentable {
    typealias NSViewType = LocalProcessTerminalView
    @State var isLoading: Bool

    private var term: LocalProcessTerminalView

    init() {
        isLoading = true
        term = LocalProcessTerminalView(frame: CGRect(x: 0, y: 0, width: 500, height: 600))
        isLoading = false
    }

    func makeNSView(context _: Context) -> LocalProcessTerminalView {
        return term
    }

    func updateNSView(_: LocalProcessTerminalView, context _: Context) {}

    mutating func startProcess(id: String, loginUser: String, isX: Bool = false) {
        isLoading = true
        term = LocalProcessTerminalView(frame: CGRect(x: 0, y: 0, width: 500, height: 600))
        term.startProcess()
        if isX {
            sendCommand(...)
        } else {
            sendCommand(...)
        }
        isLoading = false
    }

    func sendCommand(_ command: String) {
        term.send(txt: "\(command)\n")
    }

    @ViewBuilder var loadingOverlay: some View {
        if isLoading {
            ProgressView()
        }
    }
}
````

**Comment by migueldeicaza** [comment]
There is no such notification for "the send finished", you being in full control over how the implementation of send is, can determine how to invoke whatever hooks you want.

From the sounds of your questions, I think that you might be hoping that "sending" means processed, and those are two very different things, and I am afraid you are about to find yourself into a world of things that are just not as robust as you hope.   This is why scripting languages like "Expect" came to be - an entire framework to deal with this unreliable world.
